### PR TITLE
Promote WatchList feature to Beta 

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Graduates the `WatchList` feature gate to Beta for kube-apiserver and enables `WatchListClient` for KCM.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
/hold 

for testing the controller-runtime repo

File: cmd/kube-controller-manager/app/options/options.go [hunk 1]
```diff
@@ -285,6 +288,16 @@ func (s *KubeControllerManagerOptions) Flags(allControllers []string, disabledBy
 	fs := fss.FlagSet("misc")
 	fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig).")
 	fs.StringVar(&s.Generic.ClientConnection.Kubeconfig, "kubeconfig", s.Generic.ClientConnection.Kubeconfig, "Path to kubeconfig file with authorization and master location information (the master location can be overridden by the master flag).")
+
+	if !utilfeature.DefaultFeatureGate.Enabled(featuregate.Feature(clientgofeaturegate.WatchListClient)) {
+		if err := utilfeature.DefaultMutableFeatureGate.OverrideDefault(featuregate.Feature(clientgofeaturegate.WatchListClient), true); err != nil {
+			// it turns out that there are some integration tests that start multiple control plane components which
```
Comment (Jefftree) [inline cmd/kube-controller-manager/app/options/options.go:294]: This is going to be a bit problematic with compatibility version. Can I ask which tests are affected?

Comment (p0lyn0mial) [reply cmd/kube-controller-manager/app/options/options.go:294]: It was `TestComponentSecureServingAndAuth` but let me check if anything has changed.

Comment (p0lyn0mial) [reply cmd/kube-controller-manager/app/options/options.go:293]: Still failing - https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128053/pull-kubernetes-integration/1846076203430055936

**Comment by p0lyn0mial** [comment]
This PR has been tested [with controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/pull/2934) to ensure that enabling the WatchList feature produces the same output as a standard list.

[All tests](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_controller-runtime/2934/pull-controller-runtime-test/1845861009965715456) from controller-runtime passed, except for `ClientWithCache Get when getting unstructured objects should call client reader when not cached` which doesn't use the watchlist feature.

To ensure the desired feature gate (FG) was enabled, [a smoke test was created](https://github.com/kubernetes-sigs/controller-runtime/pull/2934/commits/ba7e9df356c098a5b3eb55913ad7bb86df3f9b1f). 

Additionally, the controller-runtime PR was tested against a `kube-apiserver` with the `WatchList` feature disabled. In this case, the smoke test failed, and e2e tests reported `the watchlist request ended with an error, falling back to the standard LIST semantics" err="ListOptions.meta.k8s.io \"\" is invalid: sendInitialEvents: Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled" resource="apps/v1, Resource=deployments`, as expected.

/hold cancel

**Comment by p0lyn0mial** [comment]
xref: https://github.com/kubernetes/enhancements/issues/3157

**Pull request merged**